clang/tools/clang-linker-wrapper/SYCLOffloadWrapper.cpp

```diff
@@ -0,0 +1,748 @@
+  StructType *getSyclDeviceImageTy() {
+    return StructType::create(
+        {
+            Type::getInt16Ty(C),       // Version
+            Type::getInt8Ty(C),        // OffloadKind
+            Type::getInt8Ty(C),        // Format
+            PointerType::getUnqual(C), // DeviceTargetSpec
+            PointerType::getUnqual(C), // CompileOptions
+            PointerType::getUnqual(C), // LinkOptions
+            PointerType::getUnqual(C), // ManifestStart
+            PointerType::getUnqual(C), // ManifestEnd
+            PointerType::getUnqual(C), // ImageStart
+            PointerType::getUnqual(C), // ImageEnd
+            PointerType::getUnqual(C), // EntriesBegin
+            PointerType::getUnqual(C), // EntriesEnd
+            PointerType::getUnqual(C), // PropertySetBegin
+            PointerType::getUnqual(C)  // PropertySetEnd
+        },
+        "__tgt_device_image");
+  }
+
+  // SYCL specific binary descriptor type.
+  // struct __tgt_bin_desc {
+  //   /// version of this structure - for backward compatibility;
+  //   /// all modifications which change order/type/offsets of existing fields
+  //   /// should increment the version.
+  //   uint16_t Version;
+  //   uint16_t NumDeviceImages;
+  //   __tgt_device_image *DeviceImages;
+  //   /// the offload entry table
+  //   __tgt_offload_entry *HostEntriesBegin;
+  //   __tgt_offload_entry *HostEntriesEnd;
+  // };
+  StructType *getSyclBinDescTy() {
+    return StructType::create(
+        {Type::getInt16Ty(C), Type::getInt16Ty(C), PointerType::getUnqual(C),
+         PointerType::getUnqual(C), PointerType::getUnqual(C)},
+        "__tgt_bin_desc");
+  }
+
+  Function *addDeclarationForNativeCPU(StringRef Name) {
+    FunctionType *NativeCPUFuncTy = FunctionType::get(
+        Type::getVoidTy(C),
+        {PointerType::getUnqual(C), PointerType::getUnqual(C)}, false);
+    FunctionType *NativeCPUBuiltinTy = FunctionType::get(
+        PointerType::getUnqual(C), {PointerType::getUnqual(C)}, false);
+    FunctionType *FTy;
+    if (Name.starts_with("__dpcpp_nativecpu"))
+      FTy = NativeCPUBuiltinTy;
+    else
+      FTy = NativeCPUFuncTy;
+    auto FCalle = M.getOrInsertFunction(
+        sycl::utils::addSYCLNativeCPUSuffix(Name).str(), FTy);
+    Function *F = dyn_cast<Function>(FCalle.getCallee());
+    if (F == nullptr)
+      report_fatal_error("Unexpected callee");
+    return F;
+  }
+
+  std::pair<Constant *, Constant *>
+  addDeclarationsForNativeCPU(std::optional<ArrayRef<char>> Entries) {
+    auto *NullPtr = llvm::ConstantPointerNull::get(PointerType::getUnqual(C));
+    if (!Entries)
+      return {NullPtr, NullPtr}; // TODO: test this line.
+
+    std::unique_ptr<MemoryBuffer> MB = MemoryBuffer::getMemBuffer(
+        StringRef(Entries->begin(), Entries->size()));
+    // the Native CPU PI Plug-in expects the BinaryStart field to point to an
+    // array of struct nativecpu_entry {
+    //   char *kernelname;
+    //   unsigned char *kernel_ptr;
+    // };
```

Comment:
Not only our SYCL offloading differs from what there is in a community, but we have a separate format for native CPU device. That is a one more `TODO` comment to add about unification. Note that this is not directly related to this PR, but rather a problem of native CPU device implementation: it could have been unified with the rest of the SYCL from the beginning, I guess.

Tagging @PietroGhg and @uwedolinsky for awareness

Reply:
Thank you for tagging us! We are looking into the implications of this for NativeCPU. 

Does this PR require a "Test for wrapping of NativeCPU" mentioned in the above TODO  https://github.com/intel/llvm/pull/11536#issuecomment-1782973390 or should this be done in a subsequent PR (along with potential updates to NativeCPU)?

Reply:
I guess, subsequent PR.

Reply:
> Thank you for tagging us! We are looking into the implications of this for NativeCPU.
> 
> Does this PR require a "Test for wrapping of NativeCPU" mentioned in the above TODO [#11536 (comment)](https://github.com/intel/llvm/pull/11536#issuecomment-1782973390) or should this be done in a subsequent PR (along with potential updates to NativeCPU)?

I think that subsequent PR is perfectly fine. If I understand correctly, main goal of this PR is to make us closer to getting rid of legacy `clang-offload-wrapper`. However, while doing so we should keep in mind our longer-term goal to upstream SYCL toolchain. So, any changes done to `clang-linker-wrapper` should be done with keeping that in mind, i.e. we should ask ourselves: will that be accepted in the upstream and which questions will they ask on review?